### PR TITLE
Discard empty upload_name before persisting file

### PR DESCRIPTION
**Pull request by Half-Shot**
I noticed this when trying to upload something via a bridge, which caused the following error:
```
2020-07-19 19:04:07,670 - synapse.rest.media.v1.media_repository - 158 - INFO - POST-75 - Stored local media in file '/home/will/synapse/media_store/local_content/oo/QT/gQwsNbILJTMsHDNmvtbr'
2020-07-19 19:04:07,671 - synapse.http.server - 79 - ERROR - POST-75 - Failed handle request via 'UploadResource': <XForwardedForRequest at 0x7f2d54636e20 method='POST' uri='/_matrix/media/r0/upload?filename=' clientproto='HTTP/1.1' site=8008>
Traceback (most recent call last):
  File "/home/will/synapse/env/lib/python3.8/site-packages/synapse/http/server.py", line 228, in _async_render_wrapper
    callback_return = await self._async_render(request)
  File "/home/will/synapse/env/lib/python3.8/site-packages/synapse/http/server.py", line 254, in _async_render
    callback_return = await raw_callback_return
  File "/home/will/synapse/env/lib/python3.8/site-packages/synapse/rest/media/v1/upload_resource.py", line 77, in _async_render_POST
    content_uri = await self.media_repo.create_content(
  File "/home/will/synapse/env/lib/python3.8/site-packages/synapse/rest/media/v1/media_repository.py", line 160, in create_content
    await self.store.store_local_media(
....
  File "/home/will/synapse/env/lib/python3.8/site-packages/synapse/storage/engines/postgres.py", line 31, in _disable_bytes_adapter
    raise Exception("Passing bytes to DB is disabled.")
Exception: Passing bytes to DB is disabled.
```

While this doesn't cause a crash or any malicious behaviour, it's a touch annoying that the file was persisted to storage but the DB insertion failed because some parameters were not checked. This PR adds two simple checks to fail early if the parameters are the wrong type.

### CODE REVIEW

**Comment by clokep** [comment]
Looks like linting / mypy failed on this.

File: synapse/rest/media/v1/media_repository.py
Comment (clokep) [inline synapse/rest/media/v1/media_repository.py:145]: Looks like this matches [the spec](https://matrix.org/docs/spec/client_server/r0.6.0#post-matrix-media-r0-upload) which doesn't say that the upload filename is required, it also seems that the API layer event handles this properly:

https://github.com/matrix-org/synapse/blob/5cdca53aa07f921029cb8027693095d150c37e32/synapse/rest/media/v1/upload_resource.py#L57-L64

**Comment by clokep** [comment]
I wonder if the additional checking should be doing in the rest code, not the handler?

**Comment by Half-Shot** [comment]
This was mostly written as a way to try and debug the error in the PR body. I don't really know why the rest layer didn't catch it, but it feels safer to me to catch it in the handler layer as well.

**Comment by clokep** [comment]
We talked about this briefly and the rest layer should be asserting the proper types are provided. Can you move those checks to `UploadResource._async_render_POST`? Thanks!

Although to be honest, looking at that code I don't see how bytes could have been passed through.

**Comment by richvdh** [comment]
it looks like this has dropped off the radar somehow. That said, @Half-Shot could you fix the lint failures?

File: synapse/rest/media/v1/upload_resource.py
Comment (clokep) [inline synapse/rest/media/v1/upload_resource.py:67]:
```suggestion
        if not len(upload_name):
```

**Comment by clokep** [comment]
I pushd a commit that should hopefully fix the lint issue here.

File: synapse/rest/media/v1/media_repository.py
Comment (clokep) [inline synapse/rest/media/v1/media_repository.py:150]:
```suggestion
            media_type: The content type of the file.
            upload_name: The name of the file, if provided.
```

**Review by clokep — APPROVED** [review]
LGTM! Assuming CI passes.

Sorry for the delay!

**Comment by clokep** [comment]
@Half-Shot This seemed worth finishing up so hope you don't mind I pushed some commits. Can you take a quick look and make sure it still seems like it would do what you were seeing break?

**Comment by Half-Shot** [comment]
> @Half-Shot This seemed worth finishing up so hope you don't mind I pushed some commits. Can you take a quick look and make sure it still seems like it would do what you were seeing break?

I can't approve it but I've updated the changelog to be more accurate, otherwise LGTM

**Comment by clokep** [comment]
Good call! I'll merge once the build passes.

File: changelog.d/7905.bugfix
Comment (clokep) [inline changelog.d/7905.bugfix:1]:
```suggestion
Fix a longstanding bug when storing a media file with an empty `upload_name`.
```

**Pull request merged**
